services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusService.kt

```diff
@@ -911,10 +918,67 @@ class HakemusService(
         if (updatedApplicationEntity.hanke.generated) {
             updatedApplicationEntity.hanke.nimi = request.name
         }
+        updateTormaystarkastelut(updatedApplicationEntity)
         sendApplicationNotifications(updatedApplicationEntity, originalContactUserIds, userId)
         return hakemusRepository.save(updatedApplicationEntity)
     }
 
+    /** Calculate the traffic nuisance indexes for each work area of an application. */
+    private fun updateTormaystarkastelut(hakemusEntity: HakemusEntity) {
+        if (hakemusEntity.hakemusEntityData is KaivuilmoitusEntityData) {
+            val hakemusEntityData = hakemusEntity.hakemusEntityData as KaivuilmoitusEntityData
+            if (hakemusEntityData.startTime != null && hakemusEntityData.endTime != null) {
+                hakemusEntity.hakemusEntityData =
+                    hakemusEntityData.copy(
+                        areas =
+                            hakemusEntityData.areas?.map { area ->
+                                updateTormaystarkastelutForArea(
+                                    area,
+                                    hakemusEntityData.startTime.toLocalDate(),
+                                    hakemusEntityData.endTime.toLocalDate(),
+                                )
+                            },
+                    )
+            }
+        }
+    }
+
+    private fun updateTormaystarkastelutForArea(
+        area: KaivuilmoitusAlue,
+        startDate: LocalDate,
+        endDate: LocalDate
+    ) =
+        area.copy(
+            tyoalueet =
+                area.tyoalueet.map { tyoalue ->
+                    updateTormaystarkasteluForTyoalue(
+                        tyoalue,
+                        startDate,
+                        endDate,
+                        area.kaistahaitta,
+                        area.kaistahaittojenPituus,
+                    )
```

Comment:
I'm not sure how useful the split between these two methods is. The call is almost as long as complicated as the method it's calling.

Reply:
Refactored the code a bit. Hopefully it is now more readable.